### PR TITLE
[WIP] A minimal public API for custom primitives

### PR DESCRIPTION
**Pull request by dfm**
Power users of JAX want to be able to control the behavior of some functions under all transformations. This is currently partially supported by the public APIs like `custom_jvp`, `custom_vjp`, and `custom_vmap`, and fully supported by the private APIs around `core.Primitive`. Two key limitations of the public APIs are:

1. There is no mechanism for customizing both forward and reverse mode AD, and
2. `custom_vmap` (besides being undocumented) doesn't support reverse mode AD unless combined with `custom_vjp`.

The first issue is particularly important for kernel library authors who need to provide performant AD across a range of use cases, or for researchers using training algorithms that rely on bidirectional or higher-order AD. The second is more of a usability issue than a fundamental limitation.

One place where these issues become particularly limiting is when writing opaque kernels using callbacks, `pallas_call`, or `ffi_call`. In those cases, good performance requires customizing all transformations, so many library authors reach for the private `core.Primitive` APIs. As a result, `core.Primitive` is widely used in the wild. This isn't a great state of affairs because the existing primitive APIs are brittle and not terribly user friendly, but this also has larger knock-on effects. For example, users of `core.Primitive` typically need to write "lowering" rules which require the use of `mlir` which is yet another non-public API. But, we don't currently have an alternative approach that we can point users to!

There are many ways that we could tackle this, and many documents have been written (by me and others) about possible approaches. But, my work on this has started to converge, and after the merge of "stackless", the implementation got somewhat more straightforward, so now seemed like a good time to push this forward.

In this PR, I lay out my current ideas, and comments would be very welcome. The goal here is to start with a minimal API surface that sufficiently covers the existing use cases. Then, future iterations would add some useful extensions as discussed below.

## The proposed API

From the user's perspective, this API would look similar to the other `custom_*` APIs:

```python
import jax.extend as jex

@jex.custom_primitive
def fun(x, y):
  ...

@fun.def_vmap
def fun_vmap(axis_size, is_batched, *args):
  ...
  return out, out_is_batched

@fun.def_jvp
def fun_jvp(primals_in, tangents_in):
  ...
  return primals_out, tangents_out

@partial(fun.def_transpose, linear_argnums=(0, ...))
def fun_transpose(cts_out, *nonlinear_args):
  ...
  return cts_in
```

But, unlike the existing APIs, none of these rules would be automatically populated unless explicitly defined. Therefore, unlike `custom_jvp`, for example, calling `vmap` on a custom primitive without `def_vmap` would raise an exception.

In this PR, these are the only hooks that I have implemented because they cover the vast majority of uses of `core.Primitive`, but the current implementation the return type of `custom_primitive(fun)` is an instance of `CustomPrimitive` which is a subclass of `core.Primitive`, so it is still possible to use a custom primitive in the existing registration tables, e.g.:

```python
pe.dce_rules[fun] = ...
```

## Alternative approaches

So far, the user-facing APIs for customizing JAX functions under transformations have focused on composable per-transformation APIs like `custom_jvp`, `custom_vjp`, `custom_vmap`, and `custom_transpose`. This is a nice approach because under this model users only ever need to implement the customizations that can't be automatically generated by JAX. Each of these customization APIs should be transparent under any transformations that it doesn't touch. This approach was hugely successful for `custom_jvp` and `custom_vjp`, and these functions are now widely used. However, `custom_vmap` and `custom_transpose` both remain incomplete and rarely used.

So one reasonable approach might be to try to increase the coverage of these composable transforms by finishing `custom_vmap` and `custom_transpose` and adding support for customizing both directions of AD. I've worked on this a little bit (e.g. [see this branch](https://github.com/jax-ml/jax/compare/main...dfm:jax:custom-ad-2) for the AD case), and there are some fundamental conceptual issues that we encountered when deriving the partial evaluation and transpose rules for `custom_vmap`. These difficulties might be avoided by some ongoing work to refactor JAX's tracing into "stages" where (for example) AD is always run before batching, but I think it is worth landing a functional API without requiring those changes. Furthermore, the motivating use cases discussed above (callbacks, `pallas_call`, and `ffi_call`) all typically require customizing all these transformations anyways, so the general composable approach is probably overkill for these cases.

## Future work

My goal with this PR is to keep the implementation simple and minimally support the most common uses of `core.Primitive`, but having a primitive API like this allows other UX improvements.

**A better VJP and linearization interface**

In [a previous iteration of this project](https://github.com/jax-ml/jax/compare/main...dfm:jax:custom-ad-2), I worked the infrastructure needed to add VJP and linearization hooks by deferring evaluation of the JVP rule. A version of this feature will be straightforward to integrate into this API:

```python
@fun.def_fwd
def fun_fwd(*args):
  ...
  return out, res

@fun.def_lin
def fun_lin(res, *tangents_in):
  ...
  return tangents_out

@fun.def_bwd
def fun_bwd(res, *cts_out):
  ...
  return cts_in

# OR ...
fun.def_vjp(fun_fwd, fun_bwd)
```

while continuing to also support JVPs.

I think that this is probably a more intuitive API for supporting reverse mode AD (vs. using transpose manually), but my plan is to add that in a follow up PR.

### CODE REVIEW

**Comment by jakevdp** [comment]
This is great! Have you thought about also adding `custom_lowering` so users can use a different impl depending on device?

**Comment by dfm** [comment]
> Have you thought about also adding `def_lowering` so users can use a different impl depending on device type?

Good question! I think that the approach I would advocate for instead would be to combine this with [`lax.platform_dependent`](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.platform_dependent.html#jax.lax.platform_dependent). `lax.platform_dependent` covers all of the use cases for platform-specific lowering that I can come up with, and here that would just live within the primal function:

```python
def fun_cpu(*args):
  ...

def fun_cuda(*args):
  ...

@jex.custom_primitive
def fun(*args):
  return lax.platform_dependent(*args, cpu=fun_cpu, cuda=fun_cuda)

...
```

What do you think of that? Are there use cases that you know of that would need something more specific?

**Comment by PhilipVinc** [comment]
May I voice the need for exposing some way to define custom partitioning logic on the primitive as well? 

Recently, we have found several cases where the automatic partitioning of jax 'breaks down' and fails to automatically partition some function that is vmapped, vjp-ed and jvp-ed (An example is `jnp.take_along_axis(a, indices, axis)`, where the first axis of `a` and `indices` are partitioned).

[`custom_partitioning`](https://github.com/netket/netket/pull/1932/files) is not enough here, because we also must support vmap and vjp.
We have investiated custom primitives and that would work, though it's complex to use the not-so-documented custom partitioning logic with it...

cc @inailuig who may know a bit more.

**Comment by dfm** [comment]
> May I voice the need for exposing some way to define custom partitioning logic on the primitive as well?

Good question! I don't think it would be feasible to incorporate partitioning logic into this implementation, at least not in the short term. I'd argue that partitioning is (currently) a somewhat orthogonal question since it is executed by XLA long after JAX is out of the loop. That's why `custom_partitioning` is such a complicated beast! That being said, it's a great point that we should make sure that `custom_primitive` can be composed properly with `custom_partitioning` - more on that below!

> `custom_partitioning` is not enough here, because we also must support vmap and vjp.

I think the usual approach here would be to compose `custom_parititioning` with `custom_vjp` and `custom_vmap` (which should work!):

```python
@custom_vjp
@custom_vmap
@custom_partitioning  # <- this probably needs to be on the inside!
def fun(*args):
  ...
```

Then, inside of your `vmap` rule, as long as you call `fun` again it should partition appropriately. This issue thread probably isn't the right place to go into too much detail about this, but feel free to open a new issue with more details about the problems you have encountered, because an approach like that should work fine!
